### ISSUE TITLE
Soft Constraints

### ISSUE THREAD

**Issue by pantor** [pantor/ruckig#45]
Regarding #17 and #44, we should introduce an option to interpret the kinematic constraints as "soft" ones. Then, we don't want to get within the kinematic constraints as fast as possible, but to avoid overshooting the limits.

This requires to disable the braking trajectory and to add the `UDUD` profiles in step 1. I'm not yet sure how to guarantee a solution in this case.

**Comment by Danielxyz123** [comment]
Hi Pantor,

is this feature already implemented in the latest version? (jerk minimizing profiles..?) 

thx for your great work!

**Comment by pantor** [comment]
No, the jerk minimizing profiles are independent of the soft constraints here. Thanks!

**Comment by abernasconi90** [comment]
Hello Pantor,
I am integrating your library in a project and I am facing the same behaviour of overshooting when changing `max_velocity` on the fly. 
I would ask if there are any update on this topic and if there are also some workaround at the moment?
I tried to set the `min_velocity` parameter equal to zero which can avoid backward movement but I cannot set it to higher value (to decrease more the overshoot) since it may be in conflict with `target_velocity`. Any suggestion?

**Comment by WatchWesWork** [comment]
I'll add some thoughts here.

In my mind a "soft limit" would be an intermediate velocity target that is independent of the overall velocity limit. The user would have to decide what is more important, reaching the final position while violating the intermediate velocity target (but staying in the velocity limit), or reaching the intermediate velocity target while violating the final position (robot passes the final position twice). 

In my implementation, I first check if there is a valid trajectory directly from initial acceleration and velocity to target acceleration and velocity. If this trajectory distance is less than the target distance, we can then try to reach the intermediate target velocity. If this distance is greater than the target distance a direct trajectory is not possible. We first have to make a correction to reduce the velocity until a direct valid trajectory is possible (similar to the current ruckig behavior).

Unfortunately, this does create some really nasty edge cases. For example, it is possible to have a valid trajectory that reaches the intermediate target velocity at an acceleration somewhere between 0 and max/min. In the cases where the trajectory never reaches max/min acceleration this is very difficult to solve (I've tried). 

A possible work around would be to force a 0 acceleration at the target velocity. If a solution is possible without overshoot we use it. If not we use the current ruckig behavior. 

It's also possible to create a valid trajectory with an intermediate target velocity that is lower than the target velocity. Ruckig does not allow this. Maybe there is not a use case for this trajectory.

I'm used to the CNC world where position and velocity limits are mostly pre-planned.  In these machines the DUDU profile would only be used in a (rare) velocity override situation. The control tries to reach the velocity override (intermediate velocity target), but it will not violate the pre-planned target position and velocity.
